docs/references/architecture/adr-113-txi.md

```diff
@@ -0,0 +1,97 @@
+# ADR 113: Modular transaction hashing
+
+## Changelog
+
+- 2024-02-05: First version (@melekes)
+
+## Status
+
+TBD
+
+## Context
+
+Transaction hashing in CometBFT is currently implemented using `crypto/tmhash`
+package, which itself relies on [`sha256`](https://pkg.go.dev/crypto/sha256) to calculate a transaction's hash.
+
+The hash is then used by the built-in indexer (to index this
+transaction) and by the RPC `tx` and `tx_search` endpoints, which allow users
+to search for a transaction using a hash.
+
+`tmhash` is also used to calculate various hashes in CometBFT (e.g.,
+evidence, consensus params, commit, header, partset header).
+
+The problem some application developers are facing is a mismatch between the
+internal/app representation of transactions and the one employed by CometBFT. For
+example, [Evmos](https://evmos.org/) wants transactions to be hashed using
+the [RLP][rlp].
+
+In order to be flexible, CometBFT needs to allow changing the transaction hashing algorithm
+if desired by the app developers.
+
+## Alternative Approaches
+
+None.
+
+## Decision
+
+Give app developers a way to provide their own transaction hash function.
+
+## Detailed Design
+
+Add `hashFn HashFn` option to `NewNode` in `node.go`.
+
+```go
+// HashFn defines an interface for the transaction hashing function.
+type HashFn interface {
```

Comment:
we should also have a custom `String` for formatting the hash result

Reply:
Done